pkg/genericapiserver/server/filters/timeout.go

```diff
@@ -18,92 +18,98 @@ package filters
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		ctx, ok := requestContextMapper.Get(req)
 		if !ok {
-			return time.After(globalTimeout), ""
+			responsewriters.InternalError(w, req, errors.New("no context found for request"))
 		}
 
 		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
 		if !ok {
-			return time.After(globalTimeout), ""
+			responsewriters.InternalError(w, req, errors.New("no request info found for request"))
 		}
 
 		if longRunning(req, requestInfo) {
-			return nil, ""
+			handler.ServeHTTP(w, req)
+			return
 		}
-		return time.After(globalTimeout), ""
-	}
-	return WithTimeout(handler, timeoutFunc)
+		handlerWithTimeout.ServeHTTP(w, req)
+	})
 }
 
-// WithTimeout returns an http.Handler that runs h with a timeout
-// determined by timeoutFunc. The new http.Handler calls h.ServeHTTP to handle
+// WithTimeout returns an http.Handler that runs handler with a timeout
+// The new http.Handler calls h.ServeHTTP to handle
 // each request, but if a call runs for longer than its time limit, the
-// handler responds with a 503 Service Unavailable error and the message
-// provided. (If msg is empty, a suitable default message will be sent.) After
-// the handler times out, writes by h to its http.ResponseWriter will return
-// http.ErrHandlerTimeout. If timeoutFunc returns a nil timeout channel, no
-// timeout will be enforced.
-func WithTimeout(h http.Handler, timeoutFunc func(*http.Request) (timeout <-chan time.Time, msg string)) http.Handler {
-	return &timeoutHandler{h, timeoutFunc}
-}
-
-type timeoutHandler struct {
-	handler http.Handler
-	timeout func(*http.Request) (<-chan time.Time, string)
-}
-
-func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	after, msg := t.timeout(r)
-	if after == nil {
-		t.handler.ServeHTTP(w, r)
-		return
-	}
-
-	done := make(chan struct{})
-	tw := newTimeoutWriter(w)
-	go func() {
-		t.handler.ServeHTTP(tw, r)
-		close(done)
-	}()
-	select {
-	case <-done:
-		return
-	case <-after:
-		tw.timeout(msg)
-	}
+// handler responds with a 504 StatusGatewayTimeout error.
+func WithTimeout(handler http.Handler, timeout time.Duration, requestContextMapper apirequest.RequestContextMapper, s runtime.NegotiatedSerializer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		timedout := time.After(timeout)
+		done := make(chan struct{})
+		tw := newTimeoutWriter(w, func() {
+			ctx, ok := requestContextMapper.Get(r)
+			if !ok {
+				responsewriters.InternalError(w, r, errors.New("no context found for request"))
+			}
+
+			requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+			if !ok {
+				responsewriters.InternalError(w, r, errors.New("no request info found for request"))
+			}
+
+			gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
+			if gvr.Version == "" {
```

Comment:
Yes, v1.Status is the default for all older API groups and for anything outside of the groups.